### PR TITLE
Add blobstore and refactor comments into posts with images (like a photo feed with captions). 

### PR DESCRIPTION
**Pull request by jirehcwe**
If the refactor to Comments is confusing, I've basically changed it to a FeedPost class that now includes an image url, which is populated from the blobstore image url returned by the blobstore servlet.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+import java.io.PrintWriter;
+
+@WebServlet("/doggo-servlet")
+public final class DoggoRedditServlet extends HttpServlet {
+
+  final String COMMENT_TEXT = "commentText";
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:51]: you can make these `static final`

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:115]: can we prevent these from being stored in the first place? How does this case end up happening?

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -83,22 +30,22 @@ function addGameNameToDOM(gameName)
...
     {
-      const commentsSection = document.getElementById('comments-section');
-      commentsSection.remove();
-      const commentContainer = document.getElementById('comments-container');
-      commentContainer.innerText = "No comments so far.";
+      const postsSection = document.getElementById('posts-section');
+      postsSection.innerText = "No posts so far.";
     }
+
+    console.log(postJson);
```
Comment (123cristen) [inline portfolio/src/main/webapp/script.js:48]: Usually better to remove these before submitting, though it's useful for debugging :)

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -111,6 +58,36 @@ function createListElement(text) {
   return liElement;
 }
 
+function createPostElement(postJson) {
```
Comment (123cristen) [inline portfolio/src/main/webapp/script.js:61]: You can simplify this by just creating a string  of the HTML contents, i.e.
`post.innerHTML = "<div><div><p id="username"></p><p id="timestamp"></p></div><div></div><div><p id="comment"><p></div></div>"`

`document.getElementById("timestamp").innerText = postJson.timeStamp;`
...

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -111,6 +58,36 @@ function createListElement(text) {
...
+  const post = document.createElement("div");
+
+  const postHeader = document.createElement("div");
+  post.appendChild(postHeader);
+  const usernameDisplay = document.createElement("p");
+  const timestampDisplay = document.createElement("p");
+  postHeader.appendChild(usernameDisplay);
+  postHeader.appendChild(timestampDisplay);
+
+  usernameDisplay.innerText = "user";
```
Comment (123cristen) [inline portfolio/src/main/webapp/script.js:71]: Should this be hardcoded or are you planning to get this information from the server too?

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+      posts.add(post);
+    }
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(posts));
+  }
+
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:99]: maybe rename `uploadFileAndReturnUrl`?

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 1 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:51]: done!

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/script.js:48]: Done!

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/script.js:61]: I considered this, but then it doesn't make my Post extendable or more readable, I was also considering that I would add more elements to the Post object like number of likes etc and wanted to document the structure of the post element.

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 2 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:115]: Do you mean we should handle it client side where case 1: form no image, send to pure text comment servlet and case 2: form with image sent to /blobstore-image-upload first?

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/script.js:71]: Will add in future commit! Will leave unresolved for now.

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/script.js:61]: will change to plaintext as discussed in the meeting :)

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 2 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java]: Removed this check by making the image required.

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 6 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:99]: technically this function doesn't upload the file right? it only retrieves the image url from blobstore. the uploading was done back in [BlobstoreUploadServlet.java](https://github.com/jirehcwe/google-software-product-sprint/pull/3/commits/0afa226f276e2f00d9ce8ee74b5fefa07404a52b).

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/script.js:69]: Resolved in later commit.

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+import java.io.PrintWriter;
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:46]: Consider reordering import statements - usually we put them all in alphabetical order with no new lines in between

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  static final String FEED_POST = "FeedPost";
+  static final String DOG_NAME = "dogName";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Entity postEntity = new Entity(FEED_POST);
+    long timestamp = System.currentTimeMillis();
+    String uploadURL = getUploadedFileUrlFromBlobstore(request, "image");
+
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:63]: Consider creating a FeedPost object here and storing that directly in the postEntity, that way you don't have to recreate it in the doGet function

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+      String commentText = (String) entity.getProperty(COMMENT_TEXT);
+      String imageUrl = (String) entity.getProperty(IMAGE_URL);
+      long timestamp = (long) entity.getProperty(TIMESTAMP);
+      long id = entity.getKey().getId();
+
+      FeedPost post = new FeedPost(dogName, commentText, imageUrl, timestamp);
+      posts.add(post);
+    }
+
+    Gson gson = new Gson();
```
Comment (123cristen) [inline portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:96]: make this a class member variable so you don't have to reconstruct it every time

File: portfolio/src/main/webapp/style.css [hunk 10]
```diff
@@ -17,6 +21,47 @@ p {
...
+
+#post-header {
+  margin-bottom: 30px;
+  overflow: auto; 
+  background-color: #aacfcf;
+}
+
+#post-content {
+  text-align: center;
+
```
Comment (123cristen) [inline portfolio/src/main/webapp/style.css:46]: Remove extra new lines where possible to make it easier to read :)

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 6 above)
Comment (123cristen) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java]: Ahh I misunderstood haha. Thanks for clarifying

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 7 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:46]: Done! I was grouping based on function but I realize for anyone who is looking for the first time, alphabetical is the best.

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 9 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java:96]: Done!

File: portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java (hunk 8 above)
Comment (jirehcwe) [reply portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java]: Apparently datastore doesn't accept non-primitive data types. However, I converted the FeedPost into a JSON object to store anyway.

File: portfolio/src/main/webapp/style.css (hunk 10 above)
Comment (jirehcwe) [reply portfolio/src/main/webapp/style.css:46]: Done!

**Pull request merged**